### PR TITLE
Better resigning behaviour

### LINKED ISSUE #263: Very low time limits weird passes
[timing bug.zip](https://github.com/ujh/iomrascalai/files/159097/timing.bug.zip)

set the time limit to 1 second for the entire game, Iomrascalai passes instead of moving

it resigns with a higher time limit

### PR DESCRIPTION
**Pull request by ujh**
Improve the behaviour when the game is clearly lost, i.e. resign earlier. Hopefully this will fix #263.

@iopq: This is marked as WIP as the benchmarks aren't finished, yet. Please don't merge yet. :)

### CODE REVIEW

File: src/engine/engine_impl/mod.rs [hunk 1]
```diff
@@ -107,51 +107,45 @@ impl EngineImpl {
...
-
-                    } else {
-                        best_node
-                    }
-                }
-            };
-            let win_ratio = n.win_ratio();
-            if win_ratio == 0.0 {
-                Pass(color)
-            } else if win_ratio < 0.01 && game.winner() != color {
```
Comment (ujh) [inline src/engine/engine_impl/mod.rs]: @iopq: The check for the winner (IIRC in the sense of Tromp/Taylor rules is probably what lead to the passing.

Comment (iopq) [inline src/engine/engine_impl/mod.rs]: The check for winner prevents a pass in Tromp/Taylor rules

Comment (ujh) [inline src/engine/engine_impl/mod.rs]: The pass move was only excluded when the bot was loosing under TT rules in the lines above. Wouldn't it be possible that it was the best move but that check for the winner then prevented the resigning code from firing? Especially as it's a `!=` here and a `==` in line 131.

Comment (iopq) [inline src/engine/engine_impl/mod.rs]: Ah, that's probably what the bug is, it was probably supposed to do the opposite, pass if it seems like it's winning, not when it seems like it's losing.

**Comment by ujh** [comment]
As expected this works:

```
13x13: 39.91% wins (172 games of 431, ± 4.6 at 95%, ± 4.69 at 99%)
9x9: 84.6% wins (423 games of 500, ± 3.15 at 95%, ± 3.21 at 99%)
```

**Pull request merged**